README.md
# About XenonMKV
XenonMKV is a video format conversion tool that takes MKV files and outputs them as MP4 files. It does not re-encode video, and only decodes and encodes audio as necessary.

You'll find this tool useful to converting videos for devices that support H.264 video and AAC audio, but do not understand the MKV container. Originally, XenonMKV was meant for Xbox 360 consoles, but I'm finding now that this tool is much more useful for my <a href="http://blackberry.com/playbook">BlackBerry PlayBook</a>.

# System Requirements
XenonMKV was built and tested on a standard Ubuntu 12.04 LTS installation (x86_64), but most of the utilities and requirements here are available for most popular *nix distributions. You will need at least Python 2.7 for the argparse library, and ideally Python 2.7.3 or later in the 2.x series.

## Packages to Install
You will need some supporting packages. Instructions below work on Ubuntu 12.04.

### Install All Dependencies
	sudo apt-get install mediainfo mkvtoolnix mplayer faac gpac

### Individual Package Requirements
*	mediainfo (<http://mediainfo.sourceforge.net/en/Download/Ubuntu>)

		sudo apt-get install mediainfo	
	Alternatively, add the official mediainfo PPA and install the package, which the developer suggests might be a good idea:
	
		sudo add-apt-repository ppa:shiki/mediainfo
		sudo apt-get update
		sudo apt-get install mediainfo
*	mkvtoolnix (<http://www.bunkus.org/videotools/mkvtoolnix/downloads.html>)

		sudo apt-get install mkvtoolnix
*	mplayer (<http://www.mplayerhq.hu/design7/news.html>)

		sudo apt-get install mplayer
*	faac (<http://www.audiocoding.com/downloads.html>)

		sudo apt-get install faac
* 	MP4Box (<https://sourceforge.net/projects/gpac/>)

		sudo apt-get install gpac
		
	To compile your own gpac version, follow these instructions: <http://gpac.wp.mines-telecom.fr/2011/04/20/compiling-gpac-on-ubuntu/>
	Then, run the following commands for great success (the SpiderMonkey JS library appears to be broken currently):
	
		./configure --enable-debug --use-js=no
		make
		
	You probably still won't be able to run it, though, unless you run `sudo make install`. I prefer not to overwrite system libraries, so I copy the application directly into /opt/gpac/0.5.0:
	
		sudo mkdir -p /opt/gpac/0.5.0
		sudo cp -R bin/gcc/* /opt/gpac/0.5.0
		
	Then to run:
	
		LD_LIBRARY_PATH="$LD_LIBRARY_PATH;/opt/gpac/0.5.0" /opt/gpac/0.5.0/MP4Box

## Ubuntu 10.04
As I still have a few systems around running Ubuntu 10.04, here are the changes required to make XenonMKV functional:

*	Install Python 2.7, either from source or add the appropriate PPA:

		sudo add-apt-repository ppa:fkrull/deadsnakes
		sudo apt-get update
		sudo apt-get install python2.7	
*	Install mediainfo by adding the PPA referenced above as it is not in the 10.04 package repository.
*	Run the application directly referencing Python 2.7:

		/usr/bin/python2.7 /path/to/xenonmkv.py [arguments]

# Suggested Applications
*	vlc

	VLC is highly useful for investigating video files in a GUI. You can see the expected number of audio and video tracks and generally confirm that your video output works as expected.
	
		sudo apt-get install vlc

*	mkvtoolnix-gui

	The GUI version of mkvtoolnix allows editing of MKV files. It is useful for constructing MKV files for test cases, or extracting specific content.
	
		sudo apt-get install mkvtoolnix-gui

# Usage
Basic usage with default settings:

	xenonmkv.py /path/to/file.mkv
	
To ensure your Xbox 360 will play the resulting file, at a possible expense of audio quality:

	xenonmkv.py /path/to/file.mkv --profile xbox360

To see all command line arguments:

	xenonmkv.py --help
	
For a quiet run (batch processing or in a cronjob):
	
	xenonmkv.py /path/to/file.mkv -q

# Suggestions/Caveats
*	If your files to convert aren't too large, distributions that mount `/tmp` as tmpfs (planned for Fedora 18, Ubuntu 12.10, Debian Wheezy) can show a significant speedup if you use `--scratch-dir /tmp`. Right now for future proofing, the scratch directory is set to `/var/tmp`.
*	Use `-vv` to find display debug information and output exactly what's going on during the processing stages.
*	Native multiple file support will be coming (eg: convert an entire directory of MKVs) but you can do something like this in the meantime to queue up a list:

		cd ~/mymkvdir
		for i in `ls *.mkv`; do /path/to/xenonmkv.py $i --destination ~/mymp4dir; done

*	Performance on an Intel Core i5-2500K CPU at 3.3GHz, with a 1TB Western Digital Black SATA hard drive:
	A 442MB source MKV file with h.264 video and 6-channel AC3 audio is converted into a PlayBook-compatible MP4 (same video, 2-channel AAC audio, quality q=150) in 40.6 seconds. This does not have any enhancements such as a tmpfs mount.
	You could probably get much better performance with a solid state drive, and obviously processor speed will have an impact here.

# Audio Downmixing/Re-Encoding
By default, XenonMKV tries not to have to resample, downmix or re-encode any part of the content provided. However, chances are your source files will contain AC3, DTS or MP3 audio that needs to be re-encoded. In this case, the original source audio will always be downmixed to a two channel AAC file before it is repackaged. 

If the audio track in your MKV file is already AAC, the next thing to consider is your playback device. The Xbox 360 will not play audio in an MP4 container unless it is 2-channel stereo, which is a highly stupid limitation. Other devices, like the PlayBook, will happily parse up to 5.1 channel audio. By using either the `--channels` or `--profile` settings, you can tell XenonMKV how many channels of audio are acceptable from an AAC source before it will aggressively re-encode and downmix to 2-channel stereo. 

In short, if you plan to play MP4s on your Xbox 360, definitely use the `--profile xbox360` setting to make sure that no more than two channels make it into the output file. If your device is more reasonable, the default settings should be fine. More profiles will be added as users confirm their own device capabilities.

# Known Issues
## MP4Box crash with backtrace
Certain video files, when MP4Box loads them to rejoin into an MP4 container, will throw a glibc error beginning with:
	
	*** glibc detected *** MP4Box: free(): invalid next size (fast): 0x0000000000cc8400 ***
	======= Backtrace: =========
	/lib/x86_64-linux-gnu/libc.so.6(+0x7e626)[0x7f0b09a77626]
	/usr/lib/nvidia-current/tls/libnvidia-tls.so.295.40(+0x1c01)[0x7f0b084c7c01]
			
This occurs with both nVidia proprietary and Nouveau open source drivers. The message above is displayed when using the "current" or "current-updates" versions (295.40, 295.49). When using the 173, 173-updates (173.14.35) or Nouveau open-source driver, the free() error is the same, but the backtrace is different:

	*** glibc detected *** MP4Box: free(): invalid next size (fast): 0x00000000022d2420 ***
	======= Backtrace: =========
	/lib/x86_64-linux-gnu/libc.so.6(+0x7e626)[0x7f5e9820a626]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(minf_del+0x3f)[0x7f5e9867c69f]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(mdia_del+0x25)[0x7f5e9867c395]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(trak_del+0x33)[0x7f5e98680c03]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(gf_isom_box_array_del+0x37)[0x7f5e98692c87]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(moov_del+0x58)[0x7f5e9867c9c8]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(gf_isom_box_array_del+0x37)[0x7f5e98692c87]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(gf_isom_delete_movie+0x3a)[0x7f5e9869ad5a]
	/usr/lib/x86_64-linux-gnu/libgpac.so.1(gf_isom_close+0x39)[0x7f5e9869c779]
	MP4Box[0x40d30c]
	/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed)[0x7f5e981ad76d]
	MP4Box[0x4085c1]

Sometimes this is an intermittent error, so MP4Box has a 3-retry mechanism that starts with the temporary files. At this time I'm not sure whether it's system memory or an issue with MP4Box/gpac. There were similar crashing issues in the Windows version, which is why multiple versions of MP4Box were bundled and used for fallback.
	
If you do see this issue in your own testing, please report it and include a link to the file that causes the problem if possible. You may also be able to get the file to convert by using different command line options, such as `--resume-previous --preserve-temp-files` or including or excluding `-vv`.

Relevant system information:

	$ uname -a
	Linux ubuntu 3.2.0-29-generic #46-Ubuntu SMP Fri Jul 27 17:03:23 UTC 2012 x86_64 x86_64 x86_64 GNU/Linux
	
My MP4Box version is the default from the 'gpac' Ubuntu 12.04 package, which is `0.4.5+svn3462~dfsg0-1`:

	$ MP4Box -version
	MP4Box - GPAC version 0.4.6-DEV-rev
	GPAC Copyright: (c) Jean Le Feuvre 2000-2005
	(c) ENST 2005-200X
	GPAC Configuration: --build=x86_64-linux-gnu --prefix=/usr --includedir=${prefix}/include --mandir=${prefix}/share/man --infodir=${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --libdir=${prefix}/lib/x86_64-linux-gnu --libexecdir=${prefix}/lib/x86_64-linux-gnu --disable-maintainer-mode --disable-dependency-tracking --prefix=/usr --mandir=${prefix}/share/man --libdir=lib/x86_64-linux-gnu --extra-cflags='-Wall -fPIC -DPIC -I/usr/include/mozjs -DXP_UNIX' --enable-joystick --enable-debug --disable-ssl
	Features: GPAC_HAS_JPEG GPAC_HAS_PNG 

A fresh compile of the 0.5.0 version available at <https://sourceforge.net/projects/gpac/> also appears to trigger this bug:

	MP4Box - GPAC version 0.5.0-rev4065
	GPAC Copyright: (c) Jean Le Feuvre 2000-2005
		(c) ENST 2005-200X
	GPAC Configuration:  --enable-debug
	Features: GPAC_DISABLE_3D 

The version included with Ubuntu 10.04 (`0.4.5-0.3ubuntu6`) does not appear to have issues.

# TODO
*	Add a ConfigParser instance, allowing default options to be read and stored in persistent files: <http://docs.python.org/library/configparser.html>
*	~~Add support for custom tool locations (eg: use newer MP4Box build in a custom directory, rather than the version defined in PATH)~~
*	~~Add support for picking a specific track from a multiple-track MKV file, rather than forcing defaults~~
*	Add way for user to pick tracks at runtime (get `--select-tracks` functional; this will not be a default option




XenonMKV/changelog.txt
[2011-03-01: REVISION 39 - "war59312"]

[FIXED] mp4box now uses a Oct 18, 2010 development build by default, 0.4.6 rev2125. (http://www.videohelp.com/tools/mp4box)

NOTE: MP4Box v0.4.6 rev. 2698 dated 21 Feb 2011 and MP4Box v0.4.6 rev. 2485 dated 14 Dec 2010 have a bug that breaks x264

[CHANGED] CopyRight Updated to 2011

[UPDATED] MKVToolnix v4.5.0 dated 31 Jan 2011 (http://www.bunkus.org/videotools/mkvtoolnix/)

[UPDATED] FFmpeg version SVN-r26400 dated Jan 17 2011 (Last Windows 32-bit static build from http://ffmpeg.arrozcru.org/autobuilds/)

[UPDATED] MPlayer/MEncoder with FFmpeg MT svn rev 32848 - "Intel Pentium 4 or better Multi Core (core2/xeon/etc)" Build (http://oss.netfarm.it/mplayer-win32.php)

[2010-12-16: REVISION 38 - "war59312"]

[FIXED] mkvinfo.exe was broken in build 37 (http://xenonmkv.ev98.net/forum/index.php/topic,1886.0.html)

[UPDATED] MKVToolnix v4.4.0 dated 02 Nov 2010 (http://www.bunkus.org/videotools/mkvtoolnix/)

[UPDATED] English Language File, Also Fixed A Typo In English.xml: "MP4Box_Start_20080403" should be "MP4Box_Start_20080413", That's April 13 instead of April 3rd.

[UPDATED] mp4box now uses a December 14, 2010 development build by default, 0.4.6 rev2485. (http://www.videohelp.com/tools/mp4box), Changed Tools > Options > Last Steps Text To: "Use 0.4.6: December 14, 2010 build (default)"

[2010-10-16: REVISION 37 - "war59312"]

[FIXED] another mp4box framerate issue. Mentioned here:
	http://xenonmkv.ev98.net/forum/index.php/topic,164.msg3054.html#msg3054

[FIXED] mkvinfo framerate issue. Mentioned here:
	http://xenonmkv.ev98.net/forum/index.php/topic,164.msg3135.html#msg3135

[CHANGED] "Log all tool output" is now enabled by default to help in debugging.

[CHANGED] Win32 build built using Visual Studio 2010 on Windows 7 x64, Also fixed compile time warning about unused variable and an error with mt.exe.

[CHANGED] No longer need the Manifest file for UAC to work correctly since being buit with VS 2010, it creates its own Manifest file.

[CHANGED] CopyRight Updated to 2010

[CHANGED] Open About and Options windows centered screen instead of top left.

[UPDATED] The first time the user runs XenonMKV it offers to automaticlly download and install the Nero AAC Encoder (neroaac.zip).
          XenonMKV was downloading a very old build of Nero AAC Encoder version 1.3.3.0 dated Sep 17 2008.
          XenonMKV now downlaods the latest version of Nero AAC Encoder build 1.5.4.0 dated Feb 18 2010.

[UPDATED] MKVToolnix v4.3.0 dated 05 Sep 2010 (http://www.bunkus.org/videotools/mkvtoolnix/)

[UPDATED] #ziplib version 0.86.0, built on 2010/05/25 (http://www.icsharpcode.net/opensource/sharpziplib/Download.aspx)

[UPDATED] mp4box now uses a Sep 30 2010 development build by default, 0.4.6-dev. (http://www.videohelp.com/tools/mp4box)

[UPDATED] MediaInfo v0.7.6.4 dated 2008-05-05 (note: latest useable build due to a bug that causes "Exception fired: attempted to divide by zero" in XenonMKV)
          Downloaded from https://sourceforge.net/projects/mediainfo/files/binary/mediainfo/0.7.06.4/

[UPDATED] FFmpeg version SVN-r25478 dated Oct 15 2010 (Download latest Windows 32-bit static build at http://ffmpeg.arrozcru.org/autobuilds/)

[UPDATED] AC3Filter tools 0.31b dated 10/01/2009 (http://ac3filter.net/projects/tools/releases)

[UPDATED] MPlayer/MEncoder with FFmpeg MT svn rev 32492 - "Intel Pentium 4 or better Multi Core (core2/xeon/etc)" Build (http://oss.netfarm.it/mplayer-win32.php)

[ADDED] Use system tray icon for main exe icon

[2009-03-30: REVISION 36 - "Tweaktown"]

[FIXED] mp4box framerate issues. Mentioned in this thread:
	http://xenonmkv.ev98.net/forum/index.php/topic,164.15.html

[UPDATED] mp4box now uses a February 2009 development build by default, 0.4.6-dev.

[2008-11-23: REVISION 35 - "New MKV Experience"]

[FIXED] Scratch folder location now more restrictive and resilient. If any problems come up, XenonMKV now takes the safe route.
	http://xenonmkv.ev98.net/forum/index.php/topic,189.0.html
	http://xenonmkv.ev98.net/forum/index.php/topic,173.0.html
	and similar threads

[ADDED] Option to output file names in all uppercase for LaCie drives:
	http://xenonmkv.ev98.net/forum/index.php/topic,50.msg1119.html#msg1119

[ADDED] Options button now more prominent in main window:
	http://xenonmkv.ev98.net/forum/index.php/topic,50.msg1024.html#msg1024

[ADDED] Remember window size and position when exiting application:
	http://xenonmkv.ev98.net/forum/index.php/topic,179.0.html
	Limitation: Position saving may not work as expected on leftmost or topmost monitors from primary in multi-monitor setups.
	When set to a negative top or left position in pixels, the window will snap back to default settings.
	No workaround possible at this time. Requires further testing.

[ADDED] Detection of file compatibility/reference frames before video is split.
	Limitation: if the file has multiple video tracks of different sizes, this may falsely flag a video.
	The initial detection takes the first track that contains a width/height parameter in the original MKV.
	Workaround: Disable reference frame/Xbox 360 compatibility check.
	http://xenonmkv.ev98.net/forum/index.php/topic,181.0.html

[FIXED] Custom split timeline and text box now synchronized.
	http://xenonmkv.ev98.net/forum/index.php/topic,178.0.html

[ADDED] Better error logging if the tool requested for a launch does not exist.

[ADDED] Command line support for scripting. Parameters are:
	XenonMKV.exe -inputfile X:\path\to\input.mkv -outputfolder X:\output\
	XenonMKV.exe -inputfolder X:\path\mkvs\ -outputfolder X:\output\
	Limitation: Support for this hasn't been tested too heavily. Console output isn't quite there yet.
	It will probably brick itself if you try anything too fancy or use these commands out of order.

[OUTSTANDING ISSUES]
	http://xenonmkv.ev98.net/forum/index.php/topic,158.0.html
	Track ordering: track 2 video, track 1 audio needs further investigation

[2008-09-27: REVISION 34 - "Delicious Pie"]


xenonmkv.py
#!/usr/bin/env python

# XenonMKV
# Jake Billo, jake@jakebillo.com
# https://github.com/jbillo/xenonmkv

import sys
import os

# Check for Python version before running argparse import
if sys.version_info[0] == 2 and sys.version_info[1] < 7:
	print "You need to be running at least Python 2.7 to use this application."
	print "Try running /usr/bin/python2.7 %s" % sys.argv[0]
	sys.exit(1)

import argparse
import subprocess
import logging
import traceback

from mkv_info_parser import MKVInfoParser
from file_utils import FileUtils
from track import *
from decoder import AudioDecoder
from encoder import FAACEncoder
from mp4box import MP4Box
from process_handler import ProcessHandler
from mkvfile import MKVFile

def cleanup_temp_files(args, log):
	if not args.preserve_temp_files:
		t = MKVTrack(log)
		f_utils = FileUtils(log, args)		
		f_utils.delete_temp_files(args.scratch_dir, t.get_possible_extensions())

# Log an exception with the stack trace in debug mode, and exit if it's a critical log type
def log_exception(log, source, e, log_type="critical"):
	getattr(log, log_type)(source + ": " + e.message)
	log.debug(traceback.format_exc())
	if log_type == "critical":
		sys.exit(1)

def main():
	# Main program begins
	log = logging.getLogger("xenonmkv")
	console_handler = logging.StreamHandler()
	formatter = logging.Formatter('%(asctime)s - %(name)s [%(levelname)s] %(message)s')
	console_handler.setFormatter(formatter)
	log.addHandler(console_handler)

	dependencies = ('mkvinfo', 'mediainfo', 'mkvextract', 'mplayer', 'faac', 'MP4Box')

	parser = argparse.ArgumentParser(description='Parse command line arguments for XenonMKV.')
	parser.add_argument('source_file', help='Path to the source MKV file')
	parser.add_argument('-d', '--destination', help='Directory to output the destination .mp4 file (default: current directory)', default='.')
	parser.add_argument('-sd', '--scratch-dir', help='Specify a scratch directory where temporary files should be stored (default: /var/tmp/)', default='/var/tmp/')
	parser.add_argument('-cfg', '--config-file', help='(Not yet implemented) Provide a configuration file that contains default arguments or settings for the application', default='')
	parser.add_argument("-p", '--profile', help="Select a standardized device profile for encoding. Current profile options are: xbox360, playbook", default="")

	output_group = parser.add_argument_group("Output options")
	output_group.add_argument('-q', '--quiet', help='Do not display output or progress from dependent tools', action='store_true')
	output_group.add_argument('-v', '--verbose', help='Verbose output', action='store_true')
	output_group.add_argument('-vv', '--debug', help='Highly verbose debug output', action='store_true')
	output_group.add_argument('-pf', '--print-file', help='Output filenames before and after converting', action='store_true')

	video_group = parser.add_argument_group("Video options", "Options for processing video.")
	video_group.add_argument('-nrp', '--no-round-par', help='When processing video, do not round pixel aspect ratio from 0.98 to 1.01 to 1:1.', action='store_true')
	video_group.add_argument('-irf', '--ignore-reference-frames', help='If the source video has too many reference frames to play on low-powered devices (Xbox, PlayBook), continue converting anyway', action='store_true')

	audio_group = parser.add_argument_group("Audio options", "Select custom audio decoding and encoding options.")
	audio_group.add_argument('-c', '--channels', help='Specify the maximum number of channels that are acceptable in the output file. Certain devices (Xbox) will not play audio with more than two channels. If the audio needs to be re-encoded at all, it will be downmixed to two channels only. Possible values for this option are 2 (stereo); 4 (surround); 5.1 or 6 (full 5.1); 7.1 or 8 (full 7.1 audio). For more details, view the README file.', default="6")
	audio_group.add_argument('-fq', '--faac-quality', help='Quality setting for FAAC when encoding WAV files to AAC. Defaults to 150 (see http://wiki.hydrogenaudio.org/index.php?title=FAAC)', default=150)

	proc_group = parser.add_argument_group("File and processing options", "These options determine how XenonMKV processes files and their contents.")
	proc_group.add_argument('-st', '--select-tracks', help="(Not yet implemented) If there are multiple tracks in the MKV file, prompt to select which ones will be used. By default, the last video and tracks flagged as 'default' in the MKV file will be used.", action='store_true')
	proc_group.add_argument('-vt', '--video-track', help="Use the specified video track. If not present in the file, the default track will be used.", type=int)
	proc_group.add_argument('-at', '--audio-track', help="Use the specified audio track. If not present in the file, the default track will be used.", type=int)
	proc_group.add_argument('-rp', '--resume-previous', help='Resume a previous run (do not recreate files if they already exist). Useful for debugging quickly if a conversion has already partially succeeded.', action='store_true')
	proc_group.add_argument('-n', '--name', help='Specify a name for the final MP4 container. Defaults to the original file name.', default="")
	proc_group.add_argument('-preserve', '--preserve-temp-files', help="Preserve temporary files on the filesystem rather than deleting them at the end of each run.", action='store_true')
	proc_group.add_argument("-eS", "--error-filesize", help="Stop processing this file if it is over 4GiB. Files of this size will not be processed correctly by some devices such as the Xbox 360, and they will not save correctly to FAT32-formatted storage. By default, you will only see a warning message, and processing will continue.", action="store_true")
	proc_group.add_argument('--mp4box-retries', help="Set the number of retry attempts for MP4Box to attempt to create a file (default: 3)", default=3, type=int)

	dep_group = parser.add_argument_group("Custom paths", "Set custom paths for the utilities used by XenonMKV.")
	for dependency in dependencies:
		dep_group.add_argument("--" + dependency.lower() + "-path", help="Set a custom complete path for the " + dependency + " tool. Any library under that path will also be loaded.")

	if len(sys.argv) < 2:
		parser.print_help()
		sys.exit(1)

	args = parser.parse_args()

	# Depending on the arguments, set the logging level appropriately.
	if args.quiet:
		log.setLevel(logging.ERROR)
	elif args.debug:
		log.setLevel(logging.DEBUG)
		log.debug("Using debug/highly verbose mode output")
	elif args.verbose:
		log.setLevel(logging.INFO)
	
	# Check for 5.1/7.1 audio with the channels setting
	if args.channels == "5.1":
		args.channels = "6"
	elif args.channels == "7.1":
		args.channels = "8"
	if args.channels not in ('2', '4', '6', '8'):
		log.warning("An invalid number of channels was specified. Falling back to 2-channel stereo audio.")
		args.channels = "2"
	
	if args.profile:
		if args.profile == "xbox360":
			args.channels = "2"
			args.error_filesize = True
		elif args.profile == "playbook":
			args.channels = "6"
			args.error_filesize = False
		else:
			log.warning("Unrecognized device profile %s" % args.profile)
			args.profile = ""

	log.debug("Starting XenonMKV")

	# Check if we have a full file path or are just specifying a file
	if os.sep not in args.source_file:
		log.debug("Ensuring that we have a complete path to %s" % args.source_file)
		args.source_file = os.path.join(os.getcwd(), args.source_file)
		log.debug("%s will be used to reference the original MKV file" % args.source_file)
	
	# Always ensure destination path ends with a slash
	if not args.destination.endswith(os.sep):
		args.destination += os.sep

	if not args.scratch_dir.endswith(os.sep):
		args.scratch_dir += os.sep

	# Initialize file utilities
	f_utils = FileUtils(log, args)

	# Check if all dependent applications are installed and available in PATH, or if they are specified
	# If so, store them in args.tool_paths so all classes have access to them as needed
	(args.tool_paths, args.library_paths) = f_utils.check_dependencies(dependencies)

	# Check if source file exists and is an appropriate size
	try: 
		f_utils.check_source_file(args.source_file)
	except IOError as e:
		log_exception(log, "check_source_file", e)
		
	source_basename = os.path.basename(args.source_file)
	source_noext = source_basename[0:source_basename.rindex(".")]

	if not args.name:
		args.name = source_noext
		log.debug("Using '%s' as final container name" % args.name)	

	# Check if destination directory exists
	try:
		f_utils.check_dest_dir(args.destination)
	except IOError as e:
		log_exception(log, "check_dest_dir", e)

	log.info("Loading source file %s " % args.source_file)

	if args.print_file:
		print "Processing: %s" % args.source_file

	try:
		to_convert = MKVFile(args.source_file, log, args)
		to_convert.get_mkvinfo()
	except Exception as e:
		cleanup_temp_files(args, log)
		log_exception(log, "get_mkvinfo", e)

	# If the user knows which A/V tracks they want, set them. MKVFile will not overwrite them.
	try: 
		if args.video_track:
			to_convert.set_video_track(args.video_track)
	except Exception as e:
		log_exception(log, "set_video_track", e, "warning")
	
	try:
		if args.audio_track:
			to_convert.set_audio_track(args.audio_track)	
	except Exception as e:
		log_exception(log, "set_audio_track", e, "warning")

	try: 
		# Check for multiple tracks
		if to_convert.has_multiple_av_tracks():
			log.debug("Source file %s has multiple audio and/or video tracks" % args.source_file)
			if args.select_tracks:
				# TODO: Add selector for tracks
				# Handle this scenario: prompt the user to select specific tracks,
				# or read command line arguments, or something.
				log.warning("Multiple track selection support is not yet implemented. Using default tracks")
				to_convert.set_default_av_tracks()
			else:
				log.debug("Selecting default audio and video tracks")
				to_convert.set_default_av_tracks()
		else:
			# Pick default (or only) audio/video tracks
			log.debug("Source file %s has 1 audio and 1 video track; using these" % args.source_file)
			to_convert.set_default_av_tracks()
	except Exception as e:
		cleanup_temp_files(args, log)
		log_exception(log, "set_default_av_tracks", e)		

	# Next phase: Extract MKV files to scratch directory
	try:
		(video_file, audio_file) = to_convert.extract_mkv()
	except Exception as e:
		cleanup_temp_files(args, log)
		log_exception(log, "extract_mkv", e)

	# If needed, hex edit the video file to make it compliant with a lower h264 profile level
	if video_file.endswith(".h264"):
		f_utils.hex_edit_video_file(video_file)

	# Detect which audio codec is in place and dump audio to WAV accordingly
	if to_convert.get_audio_track().needs_recode:
		log.debug("Audio track %s needs to be re-encoded" % audio_file)
		audio_dec = AudioDecoder(audio_file, log, args)
		audio_dec.decode()

		# Once audio has been decoded to a WAV, use the FAAC application to encode it to .aac
		faac_enc = FAACEncoder(os.path.join(args.scratch_dir, "audiodump.wav"), log, args)
		faac_enc.encode()
		encoded_audio = os.path.join(args.scratch_dir, "audiodump.aac")
	else:
		# Bypass this whole encoding shenanigans and just reference the already-valid audio file
		encoded_audio = audio_file

	# Now, throw things back together into a .mp4 container with MP4Box.
	video_track = to_convert.get_video_track()
	mp4box = MP4Box(video_file, encoded_audio, video_track.frame_rate, video_track.pixel_ar, args, log)
	try:
		mp4box.package()
	except Exception as e:
		cleanup_temp_files(args, log)
		log_exception(log, "package", e)

	# Move the file to the destination directory with the original name
	dest_path = os.path.join(args.destination, source_noext + ".mp4")
	os.rename(os.path.join(args.scratch_dir, "output.mp4"), dest_path)

	log.info("Processing of %s complete; file saved as %s" % (args.source_file, dest_path))

	# Delete temporary files if possible
	cleanup_temp_files(args, log)

	log.debug("XenonMKV completed processing")
	if args.print_file:
		print "Completed: %s" % dest_path	
	


if __name__ == "__main__":
	main()

